### ISSUE TITLE
Design: Triggered Prod Deployments

### ISSUE THREAD

**Issue by DanVanAtta** [triplea-game/triplea#5602]
Extracting this conversation from a PR: https://github.com/triplea-game/triplea/pull/5586#discussion_r350601067

At question is how do we define the version that will be deployed in production, how to indicates and store that, and how do we trigger the deployment?

In the PR I'm essentially we keep that version number stored in the travis config. We would run the production deployment at the end of every merged build. Ansible is designed to be idempotent, it's also not a bad practice to run it somewhat frequently so that any by-hand edits on the server are discarded and we get back into a known state (we should avoid updating anything by hand in production or on any of the servers and instead favor the 'infrastructure-as-code' mindset).

### Pros/Cons to this approach

Pro:
 - version changes are in history
 - because version changes are in history it serves as a reference for how to change prod version
 - version changes are notified via PR email notifications
 - version changes can be suggested by anyone
 - when version does not change, we are still doing a production update to for example update packages, update/install new SSH keys. We can otherwise still control/modify the production infrastructure without changing the running applications. Through this too we also avoid having to keep track of multiple branches and back-port changes from one branch to another
- prerelease deployment will happen before production deployment, so we can get some verification before actually updating production
- deployments are automatic, nobody needs to install ansible and/or remember the commands
- we get a history of the deployment output in travis
- we can re-trigger deployments and deployment steps in travis by restarting the build job

Cons:
- we do need a build to execute the deployment, this introduces a delay
- hardcoded value kinda buried in the travis config, would be nice to make this more front and center

What other options do we have?

### Put prod version in a config file
This is just extracting the version number from travis. Only drawback is we add some minor parsing logic to travis to grab this file number, we perhaps introduce some indirection. 

Pro: we do not need to modify the travis configuration file when updating production version number.

### Travis Env Variable
Drawbacks here are that we do not get tracking of when the prod version updates

### Use a special production branch or tag

We could update the prod deploy if statement to check for a specific branch or tag prefix. The drawback here is that new updates to the infrastructure are not deployed out. We'd have to maintain master and the production branch(es) to have the latest that we want updated.

We also run into the issue of infrequent deployments, where we would be sending a lot of queued up changes to production all at once whenever we backport the latest master to prod or cut a new prod.

## Additional Considerations

The fact we will deploy latest config and latest binaries to prerelease is good, but we will have gap where we will not verify that the latest config and the production artifacts would work together. We may want to add a stage that will do a deploy to localhost to simulate the production deployment. We can also consider having this be a linode so we can simulate the incremental changes DB changes. 

To summarize the proposed two stages:
- 1. prod simulation run on docker (the DB updates would be executed from scratch)
- 2. prod simulation run on linode (DB updates are cumulative, so we would simulate the incremental update when we do update version number).

**Comment by RoiEXLab** [comment]
I think having a single source of truth for the current stable version really is key. Currently the source of truth for the website is the latest non-prerelease entry here on github.

In don't know if changing the prerelease status of a release does currently trigger a webhook or potentially a GH action, otherwise we could easily keep everything as is, but add an automated workflow, so that the new version spreads everywhere.

**Comment by DanVanAtta** [comment]
The single-source should be well defined, I agree.

At question though I think is whether to have the capability to deploy backend independent of releasing a new front-end version.

Decoupling the two has some benefits:
- backend could be updated much more rapidly than front-end. We don't want our users to have to keep pace, we're striving for a situation where we can deploy frequently, perhaps once a week or even bi-daily if we have the capacity.
- frontend is risky to change, requires heavy process with full regression testing and thorough QA. Incurring that anytime we want to change the backend is a big overhead. It'll be a great day when most of the front-end is well tested, until then the process is very heavy. Having to wait for front-end to be regression tested could get painful
- there is a lot more backend possibilities to update and compatibility can be kept. We can break front-end compatibility but still be doing backend updates that maintain compatibility.  I expect we'll eventually build up new features and then have front-end updates that will be compatible and recommended, but optional.
